### PR TITLE
SyncLock: Don't recalc half/double multiplier on every callback

### PR DESCRIPTION
**Pull request by ywwg**
This PR builds on #3698. It contains some refactoring to make the functions clearer.

### CODE REVIEW

File: src/engine/sync/synccontrol.cpp [hunk 1]
```diff
@@ -306,23 +322,15 @@ void SyncControl::updateTargetBeatDistance() {
...
+void SyncControl::updateInstantaneousBpm(double bpm) {
     // Adjust the incoming bpm by the multiplier.
-    m_pBpmControl->setInstantaneousBpm(bpm * m_masterBpmAdjustFactor);
-}
-
-void SyncControl::reportTrackPosition(double fractionalPlaypos) {
-    // If we're close to the end, and master, disable master so we don't stop
-    // the party.
-    if (isMaster(getSyncMode()) &&
-            fractionalPlaypos > kTrackPositionMasterHandoff) {
```
Comment (Holzhaus) [inline src/engine/sync/synccontrol.cpp:323]: `kTrackPositionMasterHandoff` is now unused, please remove.

Comment (ywwg) [reply src/engine/sync/synccontrol.cpp:323]: removed in the upstream #3698

Comment (ywwg) [reply src/engine/sync/synccontrol.cpp:323]: (merged)

**Comment by ywwg** [comment]
This patch is ready for review except that it builds on #3698 and there's no way to tell github to show the diff to that patch instead of main

**Comment by ywwg** [comment]
rebased / updated etc! ready for review

**Comment by daschuer** [comment]
Did you forget to push? I assume you want to make a single commit on top of main.

**Comment by daschuer** [comment]
Which code line contains the fix? Maybe it is also to have a single commit containing the fix only.

File: src/engine/sync/enginesync.cpp [hunk 2]
```diff
@@ -399,11 +407,13 @@ void EngineSync::requestBpmUpdate(Syncable* pSyncable, double bpm) {
     }
 
     if (mbaseBpm != 0.0) {
-        // resync to current master
-        pSyncable->setMasterParams(beatDistance, mbaseBpm, mbpm);
+        // update from current master
+        pSyncable->updateMasterBeatDistance(beatDistance);
+        pSyncable->updateMasterBpm(mbpm);
```
Comment (ywwg) [inline src/engine/sync/enginesync.cpp:412]: this is a part of the fix -- we just update the beat distance and bpm, NOT basebpm

**Comment by ywwg** [comment]
rebased and squashed. Sorry to mix the refactor with the fix.

**Comment by daschuer** [comment]
Is this still [WIP] according to the title?

**Review by daschuer — CHANGES_REQUESTED** [review]
I think there is a remaining issue: When I sync to a non const tempo track at master in a slow start region with 100 BPM a cont track is synced to it it follows the track lets say to 150 BPM. 
When I change now the explicite master to the const track, the original master haves the tempo. 
That was at least surprising. I think we should also in this case not recalc the multiplier.

**Comment by ywwg** [comment]
I am finding major half/double issues in the code in general :(.  I will try to fix those first.

**Comment by ywwg** [comment]
This now is a followon to #3899

**Comment by ywwg** [comment]
I'll split out the fix

**Comment by ywwg** [comment]
ok should be good now

**Comment by daschuer** [comment]
This has a failing test now.

**Comment by daschuer** [comment]
I can confirm that the original issue is fixed, now random pitch change when a track is playing.

I have only discovered this weird case: 

When I have a playing track in deck A and it has a tempo change in this case form 107 to 100 and the follower is cued to take over at 160 it sits at 107 than suddenly changes to 200.
This a constructed edge case, but the result is surprising.  

Of cause this does not happen when the internal clock takes over.

**Comment by ywwg** [comment]
fixed broken test (force push)

**Comment by daschuer** [comment]
Don't rebase PRs under review. I can now review and test the whole OR again, because I cannot see what has changed and fixed the test.

Do you have the original commits you .git folder?
Maybe you can re-create them or at least point me to your changes.

**Comment by daschuer** [comment]
Fortunately I had the original testing branch in my .git folder.
The fix was just: 

```
index cfe50230b1..fc51115364 100644
--- a/src/test/enginesynctest.cpp
+++ b/src/test/enginesynctest.cpp
@@ -1839,7 +1839,7 @@ TEST_F(EngineSyncTest, HalfDoubleThenPlay) {
     ProcessBuffer();
     pButtonSyncEnabled2->slotSet(1.0);
     pButtonSyncEnabled1->slotSet(1.0);
-    EXPECT_DOUBLE_EQ(0.5,
+    EXPECT_DOUBLE_EQ(1.0,
             m_pChannel1->getEngineBuffer()
                     ->m_pSyncControl->m_masterBpmAdjustFactor);
     EXPECT_DOUBLE_EQ(1.0,
```
Next time it would be much less hassle if you just commit that fix. Easy to review and now business logic touched.

**Review by daschuer — APPROVED** [review]
